pkg/controller/servicebindingrequest/servicebinder.go

```diff
@@ -236,40 +241,102 @@ func (b *serviceBinder) onError(
 	return requeueOnNotFound(err, requeueAfter)
 }
 
+// isApplicationSelectorEmpty returns true if applicationSelector is not declared in
+// the Service Binding Request.
+func isApplicationSelectorEmpty(
+	application v1alpha1.ApplicationSelector,
+) bool {
+	var emptyApplication v1alpha1.ApplicationSelector
+	if application == emptyApplication ||
+		application.ResourceRef == "" &&
+			application.LabelSelector.MatchLabels == nil {
+		return true
+	}
+	return false
+}
+
+// handleApplicationError handles scenarios when:
+// 1. applicationSelector not declared in the Service Binding Request
+// 2. application not found
+func (b *serviceBinder) handleApplicationError(reason string, applicationError error) (reconcile.Result, error) {
+	conditionsv1.SetStatusCondition(&b.sbr.Status.Conditions, conditionsv1.Condition{
+		Type:   CollectionReady,
+		Status: corev1.ConditionTrue,
+	})
+	conditionsv1.SetStatusCondition(&b.sbr.Status.Conditions, conditionsv1.Condition{
+		Type:    InjectionReady,
+		Status:  corev1.ConditionFalse,
+		Reason:  reason,
+		Message: applicationError.Error(),
+	})
+
+	// updating status of request instance
+	sbr, err := b.updateStatusServiceBindingRequest(b.sbr, &b.sbr.Status)
+	if err != nil {
+		return requeueOnConflict(err)
+	}
+
+	// appending finalizer, should be later removed upon resource deletion
+	sbr.SetFinalizers(append(removeStringSlice(b.sbr.GetFinalizers(), finalizer), finalizer))
+	if _, err = b.updateServiceBindingRequest(sbr); err != nil {
+		return noRequeue(err)
+	}
+
+	b.logger.Info(applicationError.Error())
+
+	if errors.Is(applicationError, errApplicationNotFound) {
+		sbr.SetFinalizers(nil)
+		if _, err = b.updateServiceBindingRequest(sbr); err != nil {
+			return noRequeue(err)
+		}
+		return requeueOnNotFound(errApplicationNotFound, requeueAfter)
+	}
+
+	return done()
+}
+
 // bind configures binding between the Service Binding Request and its related objects.
 func (b *serviceBinder) bind() (reconcile.Result, error) {
-	sbrStatus := b.sbr.Status.DeepCopy()
-
 	b.logger.Info("Saving data on intermediary secret...")
 
 	secretObj, err := b.secret.createOrUpdate(b.envVars, b.sbr.AsOwnerReference())
 	if err != nil {
 		b.logger.Error(err, "On saving secret data..")
-		return b.onError(err, b.sbr, sbrStatus, nil)
+		return b.onError(err, b.sbr, &b.sbr.Status, nil)
 	}
-	sbrStatus.Secret = secretObj.GetName()
+	b.sbr.Status.Secret = secretObj.GetName()
 
+	if isApplicationSelectorEmpty(b.sbr.Spec.ApplicationSelector) {
+		return b.handleApplicationError(EmptyApplicationSelectorReason, errEmptyApplicationSelector)
+	}
 	updatedObjects, err := b.binder.bind()
 	if err != nil {
 		b.logger.Error(err, "On binding application.")
-		return b.onError(err, b.sbr, sbrStatus, updatedObjects)
+		if errors.Is(err, errApplicationNotFound) {
```

Comment:
I believe @qibobo has mentioned something about this assertion not working as expected, could you double check whether this works as expected?

Reply:
Checked. This scenario is working as expected